### ISSUE TITLE
Core: make geometry variable at runtime

### ISSUE THREAD

**Issue by GPMueller** [spirit-code/spirit#102]
Three things must be considered:
- Geometry
- Hamiltonian_Anisotropic
- Hamiltonian_Isotropic (less important)

**Comment by GPMueller** [comment]
Requires: #101 and #117

**Comment by GPMueller** [comment]
To cite @SpiritSuperUser:

It might be useful, to make the lattice dimensions adjustable on the fly by adding/deleting a given number of rows in each direction without changing the other parts of the lattice. with this, is would be easier to analyze the effect of boundary conditions and lattice dimensions on a structure.

if one adds rows, one might have to think about how and where to initialize the new ones.
My suggestion:
either

- ferromagnet in a definable direction
- copy of the neighboring already existing row and define number of rows to be added on both sides of each direction
- add a "centered" check box which averages when activated automatically the two input values to add the same amount of rows in each direction

**Comment by GPMueller** [comment]
Some considerations:
- the geometry's n_cells and spin positions need to be updated
- the spin_system's and the state-clipboard's spins need to be updated
- all other NOS-sized arrays need to be resized (for others the previous data should not matter)
- the Hamiltonian needs to check if all pairs/neighbours lie within the new system size (esp. DDI)

Probably it makes sense to only allow this change for the whole collection at once, as otherwise checks for fitting NOS between copy/paste, gneb etc. would have to be added.

In order to update spins we need to adequately copy over existing data to a larger (or smaller) array. The spin_positions can be easily re-generated.

**Comment by GPMueller** [comment]
Started work on this on `feature-variable-geometry` with 2f9690b19ce509dddf71fb679a678b16c98753b2.

**Comment by GPMueller** [comment]
#160 should be resolved in order for variable geometry to function properly.

**Comment by GPMueller** [comment]
With 3f2eb1c6f89499dcfdfb401fc2b780919476cc9a on `feature-refactor-geometry`, work on this has resumed. The `Geometry` construction mechanism has been significantly improved.

**Comment by GPMueller** [comment]
With f79f5d0687bf3a0f4aca9c0573462e7034ef5fd0 on `feature-refactor-geometry`, the `Geometry` has been made runtime-variable in a crash-free way. Both `Geometry` and `Hamiltonian` are now far less dependent on system size and instead mostly on the basis cell.

**What to do when the basis cell is changed?** Especially the `Hamiltonian` would have to be treated accordingly - pairs would no longer be valid, but could be re-calculated if information is available in form of neighbour shells.

**Comment by GPMueller** [comment]
The ground work was merged into `develop` with f36fea8e670a895ef55228c2818dc7822cbf2810.

**TODO:**
- [x] API function to change n_cells
- [x] Refactor parser and the Geometry class to contain more useful information and easier handle changes during runtime (see also #117)
- [ ] API function to change the bravais vectors/lattice
- [ ] API function to change the lattice constant
- [ ] Add some more bravais lattice types
- [ ] Function to re-package `vectorfield`s, `scalarfield`s and `intfield`s into new geometry, without changing the geometric ordering of the data (i.e. change the ordering in `the std::vector` accordingly)
- [ ] API functions to change the basis cell (either need to call other API functions to delete all interactions or notify the user somehow)

**Comment by GPMueller** [comment]
Implemented function to re-distribute vector-fields on new Geometries with ef7028444e91c5e36df41616a65e7251c5d1007a.

**Comment by GPMueller** [comment]
I don't know when, but `Geometry_Set_Cell_Atoms` has been implemented and available bravais lattice types are:
- SC
- FCC
- BCC
- Hex2D (60deg)
- Hed2D (120deg)

I believe the QT UI can soon add the corresponding tab permanently and remove the "might break the code" remark (also as features are now more tested since c3422bcad2a42798e095b9e68f7565398b23ecb2).
